compiler/modules/bir/verify.bal

```diff
@@ -277,8 +277,38 @@ function verifyConvertToFloatInsn(VerifyContext vc, ConvertToFloatInsn insn) ret
     }
 }
 
+function verifyCompareOperandTypes(VerifyContext vc, t:UniformTypeBitSet expectType, CompareInsn insn, "memberType"?memberType=()) returns err:Semantic? {
+    foreach var operand in insn.operands {
+        t:SemType operandType;
+        if operand is Register {
+            operandType = operand.semType;
+        }
+        else {
+            operandType = t:constBasicType(operand);
+        }
+        if memberType is () {
+            if !t:isSubtypeSimple(operandType, expectType) {
+                return vc.err(`operand of ${insn.op} does not match order type`);
+            }
+        }
+        else {
+            t:UniformTypeBitSet? operandMemberTy = t:simpleArrayMemberType(vc.typeEnv(), operandType);
+            if operandMemberTy is t:UniformTypeBitSet {
+                t:UniformTypeBitSet optExpectedType = t:uniformTypeUnion(expectType | (1 << t:UT_NIL));
+                if !t:isSubtypeSimple(operandMemberTy, optExpectedType) {
+                    return vc.err(`operand of ${insn.op} are not same array type`);
+                }
+            }
+            else {
+                panic err:impossible("Failed to get array member type");
+            }
+        }
+    }
+}
+
 function verifyCompare(VerifyContext vc, CompareInsn insn) returns err:Semantic? {
```

Comment:
I think we can improve this. One minor point: please order things top down with verifyCompare first followed by the helper functions.

In verifyCompare, iterate over each operand, and compute the actual type (as you do in verifyCompareOperandTypes). For each operand, then call a function to check the that the semType of the operand is compatible with the orderType; there should separate functions F and G for list and non-list, respectively. F should call G.

Reply:
Fixed